epix/run_epix.py

```diff
@@ -35,6 +36,8 @@ def main(args, return_df=False, return_wfsim_instructions=False, strax=False):
 
     # Cluster finding and clustering (convert micro_separation mm -> cm):
     inter = epix.find_cluster(inter, args['micro_separation']/10, args['micro_separation_time'])
+    if len(inter) == 0 and return_wfsim_instructions:
+        return np.array([], dtype=wfsim.instruction_dtype) # empty
```

Comment:
As I said before I would like to reduce the complexity of epix by having just a single return statement. Already now it is too in transparent in my opinion.

Reply:
In addition it still could be that your resulting array becomes empty after `res_det = epix.in_sensitive_volume(result, args['detector_config'])`. So you would have to add another if empty....

Reply:
I forget to post also here after slack discussion.
I think that earlier return is not problem. 
For special cases like which are discussed now, it contributes readability.
We can forget special case in earlier line, then keep clear the following functions.
I think we don't need to implement something for empty treatment for all functions and call them in such cases.

if it becomes empty after res_det = epix.in_sensitive_volume(result, args['detector_config']), It will capture https://github.com/XENONnT/epix/blob/53d48c8a192c654358d01f35cb8d1a2c82908369/epix/run_epix.py#L75

Reply:
This should be removed too, as it is duplicated code.

Reply:
FYI https://www.anthonysteele.co.uk/TheSingleReturnLaw.html

Reply:
FYI: https://medium.com/@huytrongnguyen1985/lessons-learnt-from-the-clean-code-robert-c-martin-cecbe2b09139 
```
"Structured Programming

Dijkstra said that every function, and every block within a function, should have one entry and one exit. Following these rules means that there should only be one return statement in a function, (...)"
```

Reply:
Hi Keita, I am no sure about what you mean in your last paragraph, but I reiterate that Daniel's point is that one that makes sense also to me. In run_epix.py you can follow the flow of the run and see the called functions which can remove entries. For these, one just needs to enable the compatibility for empty arrays. This can be tested in [this notebook](https://github.com/XENONnT/analysiscode/blob/master/FullChainSimTests/EPIX_Hands_On.ipynb), for example (not up to date with the most recent epix, but definitely good enough for our purposes). What is exactly the complexity that you are referring to? Sorry if I am forgetting something trivial.

Reply:
Thank you very much for your comment, Diego.
I still disagree or I don't understand your points.
Let me confirm the points.
1) Earlier return increases code complexity.
2) Epix output will be suddenly empty. We should add the treatments for every function.

As you show the flow of `run_epix.py`, it has much functions which can't accept empty arrays. It is heavy job to apply the additional treatments. In addition, it is not clear what should them return as empty array (python array?, empty dict, or empty numpy structured array with some dtype?) It depends on the contexts, thus it will increase the complexity for the codes. Moreover, I think that these issues has not been considered in this architecture level.  it can't be fixed with a patch like this PR. We need to consider re-design this epix framework. (Of course, in another PR, I hope that someone will be assigned.)

I strongly believe that we should return earlier in such clear empty case to keep it clean. Actually, I think that it works all cases which I see. I think that we should care 99% case with higher priority. Of course, if someone report the problematic inputs, I will examine them and I can apply the another treatments. This is my thought for Point 1 and 2..

Reply:
HI Keita, thanks for your considerations, but I simply disagree. Please do not miss understand me. My reply is nothing personal, but let me maybe try clarify things, how I see it and why I am so resistant in this point.  

> It is heavy job to apply the additional treatments. 

I agree, but this is something Joran and I do every day for strax and straxen. We are reviewing each week for several hours others PRs to make sure we can keep a clean and maintainable code. I am also willing to help and share the burden, but first I have to add some last things for the next large strax/straxen release.

> In addition, it is not clear what should them return as empty array.

This is quite easy to answer a function should return the same "thing", but empty. So if it is a DataFrame it is an empty dataframe if it is an awkward array it should be an empty awkward array etc. 

> It depends on the contexts, thus it will increase the complexity for the codes. 

No, because it is the same thing as the function normally would do.

> Moreover, I think that these issues has not been considered in this architecture level.  

Yes, I agree and this was for sure my mistake. Firstly, because I wrote fractions of the code, secondly I was less insensitive in the past. 

 > it can't be fixed with a patch like this PR. We need to consider re-design this epix framework.

Sure it can, but as you say it will become longer. 

>  (Of course, in another PR, I hope that someone will be assigned.)

I think we as a collaboration would do better, if we would not always expect that these "low level" tasks are always fixed by the same people. Joran, Evan and others spent each weeks hours to review PRs, fix bugs add low level features everybody benefits from without having any personal gain for our PhDs etc. 

>  I strongly believe that we should return earlier in such clear empty case to keep it clean. (...) Of course, if someone report the problematic inputs, I will examine them and I can apply the another treatments

I simply disagree here. We have already an early return with an if-else statement because we were to label. Now we would add another if-else. Someone implements a new feature and adds another if-else statement and another if-else.... and another if-else. Does this kind of code sound familiar? I would prefer not to end up with code which is very difficult to maintain, like our GEANT4 code.

Reply:
Thank you @ramirezdiego and @WenzDaniel . I made an alternative solution following your comments #40 .
Let's discuss on the codes.